parfive/downloader.py

```diff
@@ -372,26 +375,120 @@ def callback(token, future, main_pb):
+                        ranges[-1][1] = ''
+                        queue = asyncio.Queue()
+                        workers = []
+                        for _range in ranges:
+                            workers.append(
+                                asyncio.create_task(Downloader._ranged_http_producer(
+                                    session, url, chunksize, _range, timeout, queue, **kwargs
+                                ))
+                            )
+                        consumer = asyncio.create_task(
+                            Downloader._ranged_http_consumer(queue, file_pb, filepath))
+                        await asyncio.gather(*workers)
+                        await queue.join()
+                        consumer.cancel()
+                        return str(filepath)
+                    else:
+                        with open(str(filepath), 'wb') as fd:
+                            while True:
+                                chunk = await resp.content.read(chunksize)
+                                if not chunk:
+                                    # Close the file progressbar
+                                    if file_pb is not None:
+                                        file_pb.close()
 
-                                return str(filepath)
+                                    return str(filepath)
 
-                            # Write this chunk to the output file.
-                            fd.write(chunk)
+                                # Write this chunk to the output file.
+                                fd.write(chunk)
 
-                            # Update the progressbar for file
-                            if file_pb is not None:
-                                file_pb.update(chunksize)
+                                # Update the progressbar for file
+                                if file_pb is not None:
+                                    file_pb.update(chunksize)
 
         except Exception as e:
             raise FailedDownload(filepath_partial, url, e)
 
+    @staticmethod
+    async def _ranged_http_consumer(queue, file_pb, filepath):
+        """
+
+        queue: `asyncio.Queue`
+             Queue for chunks
+
+        file_pb : `tqdm.tqdm` or `False`
+            Should progress bars be displayed for each file downloaded.
+
+        filepath: `pathlib.Path`
+            Path to the which the file should be downloaded.
+        """
+        with open(filepath, 'wb') as f:
+            while True:
+                offset, chunk = await queue.get()
+                # breakpoint()
+
+                f.seek(offset)
+                f.write(chunk)
+                f.flush()
+
+                # Update the progressbar for file
+                if file_pb is not None:
+                    file_pb.update(len(chunk))
+
+                queue.task_done()
+
+    @staticmethod
+    async def _ranged_http_producer(session, url, chunksize, http_range, timeout, queue, **kwargs):
```

Comment:
I am not a great fan of these method names, can we name them after what they are doing i.e. 'download' and 'write' rather than the design pattern?